includes/notifications/controllers/class.llms.notification.controller.upcoming.payment.reminder.php

```diff
@@ -0,0 +1,426 @@
+			return new WP_Error( 'upcoming-payment-reminder-passed', __( 'Upcoming payment reminder passed', 'lifterlms' ) );
+		}
+
+		// Schedule upcoming payment reminder.
+		return as_schedule_single_action(
+			$reminder_date,
+			'llms_send_upcoming_payment_reminder_notification',
+			$action_args
+		);
+
+	}
+
+	/**
+	 * Retrieve the date to remind user before actual payment
+	 *
+	 * @since [version]
+	 *
+	 * @param LLMS_Order $order        Instance of the LLMS_Order which we'll schedule the payment reminder for.
+	 * @param integer    $payment_date Optional. The upcoming payment due date in Unix time format and UTC. Default is 0.
+	 *                                 When not provided it'll be calculated from the order.
+	 * @return WP_Error|integer Returns a WP_Error if there's no payment scheduled, otherwise the reminder date in Unix format and UTC.
+	 */
+	private function get_upcoming_payment_reminder_date( $order, $payment_date = 0 ) {
+
+		$next_payment_date = $payment_date ? $payment_date : $order->get_recurring_payment_due_date_for_scheduler();
+		if ( is_wp_error( $next_payment_date ) ) {
+			return $next_payment_date;
+		}
+
+		/**
+		 * Filters the number of days before the upcoming payment due date when to notify the customer
+		 *
+		 * @since [version]
+		 *
+		 * @param integer    $days  The number of days before the upcoming payment due date when to notify the customer.
+		 * @param LLMS_Order $order Order object.
+		 */
+		$days = apply_filters( 'llms_notification_upcoming_payment_reminder_reminder_days', $this->get_option( 'reminder_days' ), $order );
+
+		// Sanitize: makes sure it's always a negative number.
+		$days = -1 * max( 1, absint( $days ) );
+
+		/**
+		 * Filters the next upcoming payment reminder date
+		 *
+		 * @since [version]
+		 *
+		 * @param integer    $upcoming_payment_reminder_time Unix timestamp for the next payment due date.
+		 * @param LLMS_Order $order                          Order object.
+		 */
+		$upcoming_payment_reminder_time = apply_filters( 'llms_notification_upcoming_payment_reminder_reminder_date', strtotime( "{$days} day", $next_payment_date ), $order );
+
+		return $upcoming_payment_reminder_time;
+
+	}
+
+
+	/**
+	 * Retrieve arguments passed to order-related events processed by the action scheduler
+	 *
+	 * @since [version]
+	 *
+	 * @param LLMS_Order $order Instance of the LLMS_Order which we'll schedule the payment reminder for.
+	 */
+	private function get_recurring_payment_reminder_action_args( $order ) {
+		return array(
+			'order_id' => $order->get( 'id' ),
+		);
+	}
+
+	/**
+	 * Set array of additional options to be added to the notification view in the admin panel
+	 *
+	 * @since [version]
+	 *
+	 * @param string $type Type of the notification.
+	 * @return array
+	 */
+	protected function set_additional_options( $type ) {
+
+		return array(
+			array(
+				'id'                => $this->get_option_name( 'reminder_days' ),
```

Comment:
We should utilize the type passed into this so that as to allow the different notification types to have different options (like other existing options)

This also means we should have a helper method `get_reminder_days( $type )` and replace calls to `get_option( 'reminder_days' )` with that helper method that'll always pull the option for the given type

```suggestion
				'id'                => $this->get_option_name( $type . '_reminder_days' ),
```

Reply:
Mmm alright, I actually thought to have one "shared" option on purpose... but makes sense and why not.

Reply:
@thomasplevy but this means that we have to schedule two different actions...
This complicates the things a bit... do we really want this?

Reply:
Alright I think the answer is yes we really want this.

Reply:
@thomasplevy 
done in https://github.com/gocodebox/lifterlms/pull/1545/commits/0ee4aecc858af901641e24360c0db3e1d8783d52
needs more tests though - I haven't ran any manual test (except for the manual sending which doesn't imply the scheduling).